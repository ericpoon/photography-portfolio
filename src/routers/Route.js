import React from 'react';
import { Route as NativeRoute } from 'react-router-dom';

const Route = (props) => {
  const { component: Component, header: Header, ...restProps } = props;
  const RouteComponent = (componentProps) => {
    if (Header) {
      return (
        <div>
          <Header />
          <Component {...componentProps} />
        </div>
      );
    }
    return <Component {...componentProps} />;
  };

  return <NativeRoute {...restProps} component={RouteComponent} />;
};

export default Route;
