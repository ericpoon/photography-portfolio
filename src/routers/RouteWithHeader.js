import React from 'react';
import { Route } from 'react-router-dom';
import Header from '../components/Header';

const RouteWithHeader = (props) => {
  const { component: Component, ...restProps } = props;
  const ActualComponent = (componentProps) => {
    return (
      <div>
        <Header />
        <Component {...componentProps} />
      </div>
    );
  };

  return <Route {...restProps} component={ActualComponent} />;
};

export default RouteWithHeader;
