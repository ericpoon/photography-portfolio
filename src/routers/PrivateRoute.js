import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import Route from './Route';

export const PrivateRoute = (props) => {
  const {
    isAuthenticated, fallbackTo, ...restProps
  } = props;
  if (isAuthenticated) return <Route {...restProps} />;
  return <Redirect to={fallbackTo} />;
};

const mapStateToProps = state => ({
  isAuthenticated: !!state.auth.uid,
});

export default connect(mapStateToProps)(PrivateRoute);
