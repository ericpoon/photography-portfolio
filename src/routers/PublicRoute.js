import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import Route from './Route';

export const PublicRoute = (props) => {
  const {
    isAuthenticated, fallbackTo, ...restProps
  } = props;
  if (isAuthenticated) return <Redirect to={fallbackTo} />;
  return <Route {...restProps} />;
};

const mapStateToProps = state => ({
  isAuthenticated: !!state.auth.uid,
});

export default connect(mapStateToProps)(PublicRoute);
