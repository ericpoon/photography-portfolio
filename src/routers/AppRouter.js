import React from 'react';
import { Router, Switch } from 'react-router-dom';
import createHistory from 'history/createBrowserHistory';
import AdminHeader from '../components/AdminHeader';
import VisitorHeader from '../components/VisitorHeader';
import NotFoundPage from '../components/pages/NotFoundPage';
import HomePage from '../components/pages/HomePage';
import LoginPage from '../components/pages/LoginPage';
import PortfolioPage from '../components/pages/PortfolioPage';
import AdminPage from '../components/pages/AdminPage';
import PrivateRoute from './PrivateRoute';
import PublicRoute from './PublicRoute';
import Route from './Route';

export const history = createHistory();

const AppRouter = () => (
  <Router history={history}>
    <div>
      <Switch>
        <Route
          path={'/'}
          component={HomePage}
          exact
        />
        <Route
          path={'/portfolio'}
          component={PortfolioPage}
          header={VisitorHeader}
          exact
        />
        <PublicRoute
          path={'/login'}
          component={LoginPage}
          fallbackTo={'/admin'}
          exact
        />
        <PrivateRoute
          path={'/admin'}
          component={AdminPage}
          header={AdminHeader}
          fallbackTo={'/login'}
          exact
        />
        <Route component={NotFoundPage} />
      </Switch>
    </div>
  </Router>
);

export default AppRouter;
