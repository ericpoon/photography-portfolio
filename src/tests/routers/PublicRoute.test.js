import React from 'react';
import { shallow } from 'enzyme';
import { PublicRoute } from '../../routers/PublicRoute';

let wrapper,
  fallbackTo,
  props;
beforeEach(() => {
  fallbackTo = 'fallback url';
  props = { prop1: 'a prop', prop2: 'another prop' };
  wrapper = shallow(
    <PublicRoute
      {...props}
      isAuthenticated
      fallbackTo={fallbackTo}
    />);
});

describe('if authenticated', () => {
  it('should render Redirect if authenticated', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.name()).toBe('Redirect');
  });

  it('should fallback correctly if redirecting', () => {
    expect(wrapper.prop('to')).toBe(fallbackTo);
  });
});

describe('if not authenticated', () => {
  beforeEach(() => {
    wrapper.setProps({ isAuthenticated: false });
  });

  it('should render public-only route if not authenticated', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.name()).toBe('Route');
  });
  it('should pass down props to actual route if not authenticated', () => {
    expect(wrapper.props()).toMatchObject(props);
  });
});
