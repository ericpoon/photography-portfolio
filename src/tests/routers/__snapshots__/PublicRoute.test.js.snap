// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if authenticated should render Redirect if authenticated 1`] = `
<Redirect
  push={false}
  to="fallback url"
/>
`;

exports[`if not authenticated should render public-only route if not authenticated 1`] = `
<Route
  prop1="a prop"
  prop2="another prop"
/>
`;
